sky/provision/kubernetes/instance.py

```diff
@@ -333,109 +333,94 @@ def _run_function_with_retries(func: Callable,
-    shell sessions.
-    """
-    set_k8s_env_var_cmd = docker_utils.SETUP_ENV_VARS_CMD
+def pre_init(namespace: str, context: Optional[str], new_nodes: List) -> None:
+    """Pre-initialization step for SkyPilot pods.
 
-    def _set_env_vars_thread(new_pod):
-        pod_name = new_pod.metadata.name
-        logger.info(f'{"-"*20}Start: Set up env vars in pod {pod_name!r} '
-                    f'{"-"*20}')
-        runner = command_runner.KubernetesCommandRunner(
-            ((namespace, context), pod_name))
+    This step is run in the pod right after it is created and before the
+    SkyPilot runtime is setup.
 
-        def _run_env_vars_cmd():
-            rc, stdout, _ = runner.run(set_k8s_env_var_cmd,
-                                       require_outputs=True,
-                                       stream_logs=False)
-            _raise_command_running_error('set env vars', set_k8s_env_var_cmd,
-                                         pod_name, rc, stdout)
+    This step includes three key steps:
 
-        _run_function_with_retries(_run_env_vars_cmd,
-                                   f'set env vars in pod {pod_name}')
-        logger.info(f'{"-"*20}End: Set up env vars in pod {pod_name!r} '
-                    f'{"-"*20}')
+    1. Privilege check: Checks if the default user has sufficient privilege
+    to set up the kubernetes instance pod.
+    2. SSH setup: Sets up SSH for the pod instance.
+    3. Environment variable setup to populate k8s env vars in the pod.
 
-    subprocess_utils.run_in_parallel(_set_env_vars_thread, new_pods,
-                                     NUM_THREADS)
+    Make sure commands used in these methods are generic and work
+    on most base images. E.g., do not use Python, since that may not
+    be installed by default.
+
+    If you run any apt commands, be sure to check if the lock is available.
+    It is possible the `apt update` run in the pod container args may still
+    be running.
+
+    Args:
+        namespace (str): Kubernetes namespace.
+        context (Optional[str]): Kubernetes context.
+        new_nodes (List): List of new pod instances.
+
+    Raises:
+        config_lib.KubernetesError: If user privileges are insufficient or
+          setup fails.
+    """
 
+    check_apt_update_complete_cmd = (
+        'echo "Checking if apt update from container init is complete..."; '
+        'timeout_secs=600; '
+        'start_time=$(date +%s); '
+        'while ! grep -q "Fetched" /tmp/apt-update.log 2>/dev/null; do '
+        '  echo "apt update still running. Logs:"; '
+        '  cat /tmp/apt-update.log; '
+        '  current_time=$(date +%s); '
+        '  elapsed=$((current_time - start_time)); '
+        '  if [ $elapsed -ge $timeout_secs ]; then '
+        '    echo "Timed out waiting for apt update"; '
+        '    exit 1; '
+        '  fi; '
+        '  sleep 5; '
+        'done; '
+        'echo "apt update complete."; ')
```

Comment:
nit: how about move this down to keep the order of definition and the order we use them the same for slightly better readability.